### ISSUE TITLE
discover via web queues arpnip & macwalk but does not run them

### ISSUE THREAD

**Issue by inphobia** [netdisco/netdisco#669]
## Expected Behavior

discover option via webinterface should run all discover steps.

## Current Behavior

it seems it runs the discover step, then queue arpnip & macwalk, but those actions never get run?

## Possible Solution

it seems discover via netdisco-do does not have this behaviour, then again, i have not found a way to pass the parameter "with-nodes" to it....

## Steps to Reproduce (for bugs)
all tables others than users, sessions, oui & dbix_class_... were truncated at the start to ensure a clean slate.

1. in the webinterface, enter an ip adres of a router & click discover
2. this will take you to the jobqueue page, the discover job should finish quite quickly
3. when the discover job is finished, it willl queue an arpnip & macsuck job for the same device.
4. it seems these jobs never ever run
![image](https://user-images.githubusercontent.com/39005454/67810699-54bc1d00-fa9b-11e9-94ab-d4eafb58e064.png)

for this test i do had the following in my config:
```yaml
devices_only:
  - 10.6.10.3
  - 10.50.102.1
```

so, to make sure this was not causing the issue, i removed devices_only from the config, truncated everything again and restarted netdisco.

i then discovered 10.6.10.3 again via the webinterface. the result of this was:
* 10.6.10.3 gets discovered
* 10.6.10.3 queues arpnip/macsuck, these still do not get run
* all neighbours (cdp/lldp/etc) get discovered
* all of those neighbors also queue arpnip/macksucks that never get run

next up i thought perhaps my schedule setup could be the issue, which is:

```yaml
schedule:
  discoverall:
    when: '2 7,17 * * *'
  macwalk:
    when:
      min: 50
      hour: '*/4'
  arpwalk:
    when:
      min: 10
      hour: '*/4'
  nbtwalk:
    when: '22 7,17 * * *'
  expire:
    when: '30 23 * * *'
  makerancidconf: null
```

so in a crazy idea i changed this to
```yaml
schedule:
  discoverall:
    when: '* * * * *'
  macwalk:
    when: '* * * * *'
  arpwalk:
    when: '* * * * *'
  nbtwalk:
    when: '* * * * *'
  expire:
    when: '30 23 * * *'
  makerancidconf: null
```

result: all the walks run like crazy, but the original arpnip/macsucks never get dequeued.

the final thing i tried was to go to the device page & press the arpnip/macwalk buttons there. this actually resulted in more queued jobs that are stuck... (please disregard timestamps, i tried many options to try & find out the cause):
![image](https://user-images.githubusercontent.com/39005454/67812757-33a9fb00-faa0-11e9-8d63-096c745b4779.png)

finally, these is what the admin table says after yet another truncate & retry:
(ignore double quotes, and the empty fields actually are null, not empty strings)

"job"|"entered"|"started"|"finished"|"device"|"port"|"action"|"subaction"|"status"|"username"|"userip"|"log"|"debug"|"device_key"
-----|---------|---------|----------|--------|------|--------|-----------|--------|----------|--------|-----|-------|------------
"79929"|"2019-10-29 23:03:40.136469"|||"10.6.10.3"||"arpnip"||"queued"|"testdisc"|"10.50.128.186"|||
"79928"|"2019-10-29 23:03:40.134108"|||"10.6.10.3"||"macsuck"||"queued"|"testdisc"|"10.50.128.186"|||
"79927"|"2019-10-29 23:03:39.147765"|"2019-10-29 23:03:39"|"2019-10-29 23:03:40"|"10.6.10.3"||"discover"|"with-nodes"|"done"|"testdisc"|"10.50.128.186"|"Ended discover for 10.6.10.3"||

and device_skip:

"backend"|"device"|"actionset"|"deferrals"|"last_defer"
---------|--------|-----------|-----------|------------
"linux002.aquafin.be"|"10.6.10.3"|"{arpnip,macsuck,nbtstat}"|0|

## Your Environment

* Netdisco version used: 2.44.3
* SNMP::Info version used: 3.70
* postgres 10.00.9
* perl 5.26.1

## Config info (deployment.yml)

for the tests i commented out most of my config. options still configured were:
* ldap config for login
* bunch of host groups and device_auth stuff
```yaml
snmptimeout: 6000000
bulkwalk_no: 'group:slowsnmp'
domain_suffix:
  - '.aquafinad.aquafin.be'
  - '.aquafin.be'
node_freshness: 1500
expire_jobs: 5
expire_nodes: 3
expire_nodes_archive: 60
devport_vlan_limit: 250
jobs_qdepth: 100
workers:
  tasks: 6
```
and device_only + scheduler configs as mentioned earlier.

other as the domain_suffix part this config has been mostly static for over a year.

**Comment by inphobia** [comment]
and the even stranger thing is if i run:
`netdisco-do discover -d 10.6.10.3` on cli, and then use the arpnip & macsuck buttons in the webinterface the run just fine.

![image](https://user-images.githubusercontent.com/39005454/67813836-df544a80-faa2-11e9-94eb-d882d81bbecc.png)

i also tried `netdisco-do discover -d 10.6.10.3 --enqueue`  and a scheduler that ran discoverall, arpwalk & macsuck every minute, and in this case the macwalk/arpnips became stuck again:

![image](https://user-images.githubusercontent.com/39005454/67814243-d617ad80-faa3-11e9-92d7-4e8667645892.png)


it seems that if a device is added via a queueing option (either via the web interface or netdisco-do discover --enqueue) discover works but arpwalk/macsuck get queued but never run. if the device was added via netdisco-do discover everything is fine. i went through the last 30 or so commits but can't see what could cause this....

**Comment by inphobia** [comment]
it seems that for some reason all device actions get added to the device device_skip table with the enqueue option, but they get run after a restart of netdisco (thought that forced a retry of all previous deferred actions.

so, clean database again and here we go:

discover device via web, discover works rest gets queued. db output (i'll just use screenshots now):
![image](https://user-images.githubusercontent.com/39005454/67817620-5b07c480-faae-11e9-9c03-373eac94e8a0.png)
![image](https://user-images.githubusercontent.com/39005454/67817635-678c1d00-faae-11e9-8b5f-aa15e568422c.png)

then i restart both backend and web, and jobs that were queued run:
![image](https://user-images.githubusercontent.com/39005454/67817691-b33ec680-faae-11e9-974f-ad5a9e0ab6d6.png)
![image](https://user-images.githubusercontent.com/39005454/67817706-bf2a8880-faae-11e9-9ea7-6a711e5fcc93.png)


i think this could be related to https://github.com/netdisco/netdisco/issues/596

**Comment by pyro3d** [comment]
Was able to reproduce. Looks like it thinks the device doesn't have L2 or L3 capability and defers the jobs. Added a quick line to print when jobs are added to device_skip:

> [86797] 2019-10-30 03:53:07 debug is_macsuckable: 192.168.200.223 has no layer 2 capability
> [86797] 2019-10-30 03:53:07 debug is_arpnipable: 192.168.200.223 has no layer 3 capability
> [86797] 2019-10-30 03:53:07 debug Deferring App::Netdisco::DB::Result::Virtual::TastyJobs=HASH(0x7feec072dd00)->device

**Comment by pyro3d** [comment]
Okay, so what I believe is happening is that during the main worker manager loop, `jq_getsome` is run, which executes `_get_denied_actions`, and defers the tasks if needed (device isn't l2-capable; don't macsuck, etc). While normally this is fine, this occurs *during* the initial discover action, and because the device hasn't been created yet, is_arpnipable and is_macsuckable both fail, and so the actions are deferred.

**Comment by pyro3d** [comment]
The changes to is_arpnipable and is_macsuckable in https://github.com/netdisco/netdisco/commit/2897eda68481f248d61ae4e46b1f2113df0e3923
are at fault here.

**Comment by taylor5042** [comment]
I just noticed a similar issue where the device_module data was not populating for a handful of devices when using the API job queue discover action or the web gui discovery. 
It seems like the arpnip and macwalk were stuck in the queue as well but I really just spot checked one or two devices. 
When using netdisco-do the device_module data was fully discovered. 
I removed my device_identity section and then ran about 100 device discoveries via the API and they all fully discovered just like they do when using netdisco-do discover. 
I'm hoping this isn't something I will run into again once I re-enable the device_identity config.
